runtime/sam/op/traverse/queryexpr.go

```diff
@@ -1,38 +1,115 @@
 package traverse
 
 import (
+	"context"
+
 	"github.com/brimdata/super"
 	"github.com/brimdata/super/runtime"
+	"github.com/brimdata/super/runtime/sam/expr"
 	"github.com/brimdata/super/zbuf"
 )
 
-type QueryExpr struct {
+type Cache struct {
 	rctx   *runtime.Context
-	puller zbuf.Puller
+	body   zbuf.Puller
 	cached *super.Value
 }
 
-func NewQueryExpr(rctx *runtime.Context, puller zbuf.Puller) *QueryExpr {
-	return &QueryExpr{rctx: rctx, puller: puller}
+func NewCache(rctx *runtime.Context, body zbuf.Puller) *Cache {
+	return &Cache{rctx: rctx, body: body}
 }
 
-func (q *QueryExpr) Eval(this super.Value) super.Value {
-	if q.cached == nil {
-		q.cached = q.exec().Ptr()
+func (c *Cache) Eval(_ super.Value) super.Value {
+	if c.cached == nil {
+		c.cached = c.exec().Ptr()
 	}
-	return *q.cached
+	return *c.cached
 }
 
-func (q *QueryExpr) exec() super.Value {
+func (c *Cache) exec() super.Value {
 	var batches []zbuf.Batch
 	for {
-		batch, err := q.puller.Pull(false)
+		batch, err := c.body.Pull(false)
 		if err != nil {
-			return q.rctx.Sctx.NewError(err)
+			return c.rctx.Sctx.NewError(err)
 		}
 		if batch == nil {
-			return combine(q.rctx.Sctx, batches)
+			return combine(c.rctx.Sctx, batches)
 		}
 		batches = append(batches, batch)
 	}
 }
+
+// QueryExpr is a simple subquery mechanism where it has both an Eval
+// method to implement expressions and a Pull method to act as the parent
+// of a subgraph that is embedded in an expression.  Whenever eval
+// is called, it constructs a single valued batch using the passed-in
+// this, posts that batch to the embedded query, then pulls from the
+// query until eos.  When the subquery is not correlated (i.e., because
+// of a new from clause or perhaps a constant values clase) then it wraps
+// the from in a cache that always returns the same batch.  Subqueries always
```

Comment:
Nit: This looks out of date.

```suggestion
// query until eos.  Subqueries always
```